src/cca/scripts/uproxy_server.ts

```diff
@@ -56,25 +36,146 @@ export class UproxyServer implements Server {
+        }
+      }, (e) => {
+        console.warn('could not inform core of saved server', e);
+      });
+    }
+  }
+
+  // returns saved contacts, which will be empty if
+  // localStorage is unavailable or no contacts are found.
+  // throws if there is an error reading from localStorage or
+  // the stored contacts could not be parsed.
+  private loadContacts(): cloud_social_provider.SavedContacts {
+    if (!isStorageAvailable) {
+      return {
+        contacts: []
+      };
+    }
+
+    const savedContactsJson = localStorage.getItem(CONTACTS_STORAGE_KEY);
+    return savedContactsJson ? <cloud_social_provider.SavedContacts>JSON.parse(
+        savedContactsJson) : {
+      contacts: []
+    };
+  }
+
+  // merges a contact with those already saved to storage.
+  // returns true if the contact was not previously in storage.
+  // contacts are saved as a SavedContacts serialised as JSON. 
+  private saveContact(newInvite:cloud_social_provider.Invite) {
+    if (!isStorageAvailable) {
+      return true;
+    }
+
+    const savedContacts = this.loadContacts();
+
+    // merge or append, as necessary.
+    const dupeIndex = _.findIndex(savedContacts.contacts, (savedContact) => {
+      return savedContact.invite && savedContact.invite.host === newInvite.host;
     });
+
+    const newSavedContact: cloud_social_provider.SavedContact = {
+      invite: newInvite
+    };
+    if (dupeIndex > -1) {
+      savedContacts.contacts[dupeIndex] = newSavedContact;
+    } else {
+      savedContacts.contacts.push(newSavedContact);
+    }
+
+    // and save!
+    localStorage.setItem(CONTACTS_STORAGE_KEY, JSON.stringify(savedContacts));
+
+    return dupeIndex === -1;
   }
 
-  public addServer(code: AccessCode): Promise<Server> {
-    let token = parseInviteUrl(code);
-    if (!token) {
-      return Promise.reject(`Failed to parse access code: ${code}`);
+  public addServer(inviteUrl:string): Promise<void> {
+    if (!this.onServerCallback) {
+      throw new Error('must call onServer first');
     }
-    // TODO: Do I need to wait for core.login()?
+
+    // inspired from ui.ts but uproxy air only supports v2 invites
+    // which have just three fields:
+    //  - v=2
+    //  - networkName=Cloud
+    //  - networkData=<jsurl stuff that we send, as an object, to the cloud social provider>
+    // TODO: accept only cloud invites
+    const params :social.InviteTokenData = uparams(inviteUrl);
+    if (!(params || params.v || params.networkName || params.networkData)) {
+      return Promise.reject(new Error('could not decode URL'));
+    }
+
+    const cloudInvite :cloud_social_provider.Invite = JSON.parse(
+        jsurl.parse(<string>params.networkData));
+
+    // inform the core of this new or updated server and, if it's new,
+    // emit an onServer event.
+    return this.informCoreOfServer(cloudInvite).then((server) => {
+      if (this.saveContact(cloudInvite)) {
+        try {
+          this.onServerCallback(server);
```

Comment:
So generally - especially as I've been re-writing the SOCKS server - I've found callbacks less problematic, e.g. callbacks can be invoked multiple times, but right I've reverted to a Promise-based API.